### DISCUSSION TITLE
SML Protocol not completely received from Landis+Gyr ULTRAHEAT T550

### DISCUSSION THREAD

**Discussion by Marty56** [arendst/Tasmota#17258] [Q&A] [answered]
Hello,

I am try to read the energy consumption from a Landis+Gyr ULTRAHEAT T550. It seems that the SML Data is incomplete and the start is truncated.
This is my log

```
08:42:42.184 : 
08:42:42.387 : 6.26*01(00278.61*m3)6.8*01(0014.048*MWh)
08:42:42.544 : F(0)9.20(67304276)6.35(7.5*m)
08:42:42.892 : 6.6(0014.0*kW)6.6*01(0014.0*kW)6.33(000.216*m3ph)9.4(117.0*C&071.6*C)
08:42:43.254 : 6.31(0024741*h)6.32(0000000*h)9.22(R)9.6(000&67304276&0&000&67304276&0)
08:42:43.584 : 9.7(20000)6.32*01(0000000*h)6.36(01-01&00:00)6.33*01(000.216*m3ph)
08:42:43.938 : 6.8.1(0005.102*MWh)6.8.2()6.8.3()6.8.4(0001.698*MWh)6.8.5(0002.881*MWh)
08:42:44.165 : 6.8.1*01(0003.741*MWh)6.8.2*01()6.8.3*01()
08:42:44.395 : 6.8.4*01(0001.171*MWh)6.8.5*01(0001.983*MWh)
08:42:44.525 : 9.4*01(117.0*C&064.3*C)
08:42:44.729 : 6.36.1(2020-03-31)6.36.1*01(2020-03-31)
08:42:44.934 : 6.36.2(2020-03-10)6.36.2*01(2020-03-10)
08:42:45.135 : 6.36.3(2021-01-25)6.36.3*01(2021-01-25)
08:42:45.339 : 6.36.4(2022-03-10)6.36.4*01(2021-04-30)
08:42:45.662 : 6.36.5()6.36*02(01&00:00)9.36(2022-12-04&09:03:22)9.24(0.6*m3ph)
08:42:45.868 : 9.17(4)9.18(040*C)9.19(045*C)9.25(050*C)
08:42:46.186 : 9.1(0&1&0&0000&CECV&CECV&1&5.20&5.20&F&101008&1>1>04&08&0&08&:5)
08:42:46.336 : 9.2(&&)9.29()9.31(0012180*h)
08:42:46.687 : 9.0.1(00000000)9.0.2(00000000)9.34.1(000.00000*m3)9.34.2(000.00000*m3)
08:42:46.878 : 8.26.1(00000000*m3)8.26.2(00000000*m3)
08:42:47.107 : 8.26.1*01(00000000*m3)8.26.2*01(00000000*m3)
08:42:47.232 : 6.26.1()6.26.4()6.26.5()
08:42:47.475 : 6.26.1*01()6.26.4*01()6.26.5*01()0.0(67304276)
08:42:47.495 : !
```

I have compiled Tasmota with the following option
```
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif
#define SML_REPLACE_VARS
#define USE_SML_SCRIPT_CMD

```

This is my script
```
>D
scnt=0
res=0

>B
=>sensor53 r

>F
;100ms iterations
scnt+=1
switch scnt
case 10
;300 baudrate
res=sml(1 0 300)
res=sml(1 1 "00000000000000000000000")
res=sml(1 1 "0000000000000000000000")
res=sml(1 1 "2F3F210D0A")
;\?!CRLF
;case 7
res=sml(1 0 2400) 
case 6000
scnt=0
ends

>M 1
+1,3,o,16,2400,LGUH50,1
1,6.8(@1,Zählerstand,kWh,waermemenge,0
#
```

I have been using V 11.0.0.3 of Tasmota and have been playing around with the "0000000" and also the "case" values because it seems that start of the SML protocol is missing and Tasmota is too late starting to receive the bits.
While doing this, I was even able to receive the first line once containing the correct value.
```
 6.8(0019.778*MWh)6.26(00407.00*m3)9.21(67304276)
```
But this was not reproducible.

Any idea what I am doing wrong would be really appreciated.

**Comment by Mart-Muc** [comment]
https://forum.creationx.de/forum/index.php?thread/1095-d0-z%C3%A4hler-sml-auslesen-mit-tasmota/&postID=38844#post38844

**Comment by Marty56** [comment, marked as answer]
Thanks for the feedback. 
This link is where I got my script from.

Meanwhile the issue is resolve and it is working.
It was a combination of weak Wifi and tedious adjustment of the IR- Head.

Whoever invented this IR-Head standard was pretty clueless or it was intentionally made bad so no-one can use it.

**Reply by Auskenner** [reply]
Hi Marty.
I am a bit confused. I flashed an esp with the settings described here and connected it to an IR read head. Unfortunately, I don't see anything in the console when I'm using your script
Any idea what that could be?

**Comment by Marty56** [comment]
There are many reasons. Is  IR Head correctly positioned, good supply voltage, ... hard to pin down the specific cause.
In my case it was a tedious trial and error approach. The position of the IR Head was not stable due to a poor magnet and the wifi connection was also not good. 
With the script you should at least see a log.  Sorry I can not help you further.

**Comment by Stargate85** [comment]
Hi, 
I have exactly the same issue and I go honestly crazy here.

I spent last evening 2 hours moving the IR sensor millimeter by millimeter but never got the first lines. Then I started to play around with the timings (case values) and voila I get the data now.

Only problem is that the parsing doesn´t work as I always have the "special character" in front of the line in the response.Anybody an idea how I can solve that / ignore the character?
`10:38:23.648 : 6.8(0053.157*MWh)6.26(00850.41*m3)9.21(70049278)`


This is my script:
```
>D
scnt=0
res=0
1,=so2,0

>B
=>sensor53 r

>F
; 100ms iterations
scnt+=1
switch scnt
case 10
  ; Set initial baud rate to 300
  res=sml(1 0 300)
  ; Send initialization sequence
  res=sml(1 1 "00000000000000000000000")
  res=sml(1 1 "0000000000000000000000")
  res=sml(1 1 "2F3F210D0A") ; /?!CRLF
case 24
  ; Change baud rate to 2400
  res=sml(1 0 2400)
case 6000
  scnt=0
ends

>M 1
+1,3,o,16,2400,LGUH50,1
1,6.8(@1,Zählerstand,MWh,waermemenge,3
```
